container/raw/inotify_watcher.go

```diff
@@ -0,0 +1,128 @@
+// Copyright 2015 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package raw
+
+import (
+	"sync"
+
+	"golang.org/x/exp/inotify"
+)
+
+// Watcher for container-related inotify events in the cgroup hierarchy.
+//
+// Implementation is thread-safe.
+type InotifyWatcher struct {
+	// Underlying inotify watcher.
+	watcher *inotify.Watcher
+
+	// Containers being watched.
+	containersWatched map[string]struct{}
+
+	// Full cgroup paths being watched.
+	cgroupsWatched map[string]struct{}
+
+	// Lock for all datastructure access.
+	lock sync.Mutex
+}
+
+func NewInotifyWatcher() (*InotifyWatcher, error) {
+	w, err := inotify.NewWatcher()
+	if err != nil {
+		return nil, err
+	}
+
+	return &InotifyWatcher{
+		watcher:           w,
+		containersWatched: make(map[string]struct{}),
+		cgroupsWatched:    make(map[string]struct{}),
+	}, nil
+}
+
+// Add a watch to the specified directory. Returns if the container was already being watched.
+func (iw *InotifyWatcher) AddWatch(containerName, dir string) (bool, error) {
+	iw.lock.Lock()
+	defer iw.lock.Unlock()
+
+	_, alreadyWatched := iw.containersWatched[containerName]
+
+	// Register an inotify notification.
+	if _, ok := iw.cgroupsWatched[dir]; !ok {
+		err := iw.watcher.AddWatch(dir, inotify.IN_CREATE|inotify.IN_DELETE|inotify.IN_MOVE)
+		if err != nil {
+			return alreadyWatched, err
+		}
+		iw.cgroupsWatched[dir] = struct{}{}
+	}
+
+	// Record our watching of the container.
+	if !alreadyWatched {
+		iw.containersWatched[containerName] = struct{}{}
+	}
+	return alreadyWatched, nil
+}
+
+// Remove watch from the specified directory. Returns if the container was already being watched.
+func (iw *InotifyWatcher) RemoveWatch(containerName, dir string) (bool, error) {
+	iw.lock.Lock()
+	defer iw.lock.Unlock()
+
+	_, alreadyWatched := iw.containersWatched[containerName]
+
+	// Remove the inotify watch if it exists.
+	if _, ok := iw.cgroupsWatched[dir]; ok {
+		err := iw.watcher.RemoveWatch(dir)
+		if err != nil {
+			return alreadyWatched, nil
+		}
+		delete(iw.cgroupsWatched, dir)
+	}
+
+	// Record the container as no longer being watched.
+	if alreadyWatched {
+		delete(iw.containersWatched, containerName)
```

Comment:
so we delete the containername on the removal of first path. Should it be when the last path is removed?